### PR TITLE
State tracker

### PR DESCRIPTION
**Pull request by Kangz**
PTAL

### CODE REVIEW

File: src/backend/CommandBufferStateTracker.cpp [hunk 1]
```diff
@@ -15,69 +15,123 @@
...
-            1 << VALIDATION_ASPECT_PIPELINE | 1 << VALIDATION_ASPECT_BIND_GROUPS |
-            1 << VALIDATION_ASPECT_VERTEX_BUFFERS | 1 << VALIDATION_ASPECT_INDEX_BUFFER;
-        if ((requiredAspects & ~mAspects).none()) {
-            // Fast return-true path if everything is good
-            return {};
+    MaybeError CommandBufferStateTracker::GenerateAspectError(ValidationAspects aspects) {
+        ASSERT(aspects.any());
+
+        if (aspects[VALIDATION_ASPECT_INDEX_BUFFER]) {
+            DAWN_RETURN_ERROR("Missing index buffer");
```
Comment (SenorBlanco) [inline src/backend/CommandBufferStateTracker.cpp]: Is it the case that only the first failing aspect will be displayed?

File: src/backend/CommandBufferStateTracker.cpp [hunk 2]
```diff
@@ -15,69 +15,123 @@
...
+    void CommandBufferStateTracker::RecomputeLazyAspects(ValidationAspects aspects) {
+        ASSERT(mAspects[VALIDATION_ASPECT_PIPELINE]);
+        ASSERT((aspects & ~kLazyAspects).none());
+
+        if (aspects[VALIDATION_ASPECT_BIND_GROUPS]) {
+            bool matches = true;
+
+            for (uint32_t i : IterateBitSet(mLastPipelineLayout->GetBindGroupLayoutsMask())) {
+                if (mLastPipelineLayout->GetBindGroupLayout(i) != mBindgroups[i]->GetLayout()) {
+                    matches = false;
```
Comment (SenorBlanco) [inline src/backend/CommandBufferStateTracker.cpp]: Nit: correct/worthwhile to break; here?

File: src/backend/CommandBufferStateTracker.cpp (hunk 1 above)
Comment (Kangz) [reply src/backend/CommandBufferStateTracker.cpp]: Yes, we'll want to have a making great error messages but that would be between the MVP and v1 I think.

File: src/backend/CommandBufferStateTracker.cpp (hunk 2 above)
Comment (Kangz) [reply src/backend/CommandBufferStateTracker.cpp]: Done

File: src/backend/CommandBufferStateTracker.cpp (hunk 1 above)
Comment (SenorBlanco) [reply src/backend/CommandBufferStateTracker.cpp]: Acknowledged.

File: src/backend/CommandBufferStateTracker.cpp [hunk 3]
```diff
@@ -90,96 +145,25 @@ namespace backend {
...
     }
 
     void CommandBufferStateTracker::SetPipelineCommon(PipelineBase* pipeline) {
-        PipelineLayoutBase* layout = pipeline->GetLayout();
+        mLastPipelineLayout = pipeline->GetLayout();
 
         mAspects.set(VALIDATION_ASPECT_PIPELINE);
-
         mAspects.reset(VALIDATION_ASPECT_BIND_GROUPS);
         mAspects.reset(VALIDATION_ASPECT_VERTEX_BUFFERS);
```
Comment (kainino0x) [inline src/backend/CommandBufferStateTracker.cpp]: Probably an existing issue, but should this also reset `VALIDATION_ASPECT_INDEX_BUFFER`?

File: src/backend/CommandBufferStateTracker.cpp [hunk 4]
```diff
@@ -15,69 +15,124 @@
...
+    MaybeError CommandBufferStateTracker::ValidateOperation(ValidationAspects requiredAspects) {
+        // Fast return-true path if everything is good
+        ValidationAspects missingAspects = requiredAspects & ~mAspects;
+        if (missingAspects.none()) {
             return {};
         }
 
-        if (!mAspects[VALIDATION_ASPECT_PIPELINE]) {
-            DAWN_RETURN_ERROR("No active compute pipeline");
+        // Generate an error immediatly if a non-lazy aspect is missing as computing lazy aspects
```
[the remark this answers is not in the document]
Comment (Kangz) [reply src/backend/CommandBufferStateTracker.cpp]: Done

**Comment by Kangz** [comment]
Yes it does loosen some validation that seemed it shouldn't have been there in the first place.

**Pull request merged**